sourcecode/shared/src/main/scala/sourcecode/SourceContext.scala

```diff
@@ -105,8 +105,17 @@ object Pkg extends SourceCompanion[String, Pkg](new Pkg(_)){
   def impl(c: Compat.Context): c.Expr[Pkg] = Impls.enclosing[Pkg](c)(_.isPackage)
 }
 
-case class Text[T](value: T, source: String)
-object Text{
+sealed trait Text[T] {
```

Comment:
sourcecode is pulled in transitively for all downstream dependencies of fastparse, which is a lot of libraries. I suspect a binary breaking change in `Text.apply` this would be quite a headache. I think it's possible to achieve this fix without breaking binary compatibility by providing the old apply/unapply/copy/productIterator/productField/... generated by the case class machinery.

```scala
  case class Text[T] extends AnyRef with Product with Serializable {
    <caseaccessor> <paramaccessor> private[this] val value: T = _;
    <stable> <caseaccessor> <accessor> <paramaccessor> def value: T = Text.this.value;
    <caseaccessor> <paramaccessor> private[this] val source: String = _;
    <stable> <caseaccessor> <accessor> <paramaccessor> def source: String = Text.this.source;
    def <init>(value: T, source: String): Text[T] = {
      Text.super.<init>();
      ()
    };
    <synthetic> def copy[T](value: T = value, source: String = source): Text[T] = new Text[T](value, source);
    <synthetic> def copy$default$1[T]: T = Text.this.value;
    <synthetic> def copy$default$2[T]: String = Text.this.source;
    override <synthetic> def productPrefix: String = "Text";
    <synthetic> def productArity: Int = 2;
    <synthetic> def productElement(x$1: Int): Any = x$1 match {
      case 0 => Text.this.value
      case 1 => Text.this.source
      case _ => throw new IndexOutOfBoundsException(x$1.toString())
    };
    override <synthetic> def productIterator: Iterator[Any] = scala.runtime.ScalaRunTime.typedProductIterator[Any](Text.this);
    <synthetic> def canEqual(x$1: Any): Boolean = x$1.$isInstanceOf[Text[_]]();
    override <synthetic> def hashCode(): Int = scala.runtime.ScalaRunTime._hashCode(Text.this);
    override <synthetic> def toString(): String = scala.runtime.ScalaRunTime._toString(Text.this);
    override <synthetic> def equals(x$1: Any): Boolean = Text.this.eq(x$1.asInstanceOf[Object]).||(x$1 match {
  case (_: Text[T]) => true
  case _ => false
}.&&({
      <synthetic> val Text$1: Text[T] = x$1.asInstanceOf[Text[T]];
      Text.this.value.==(Text$1.value).&&(Text.this.source.==(Text$1.source)).&&(Text$1.canEqual(Text.this))
    }))
  };
  <synthetic> object Text extends AnyRef with Serializable {
    def <init>(): Text.type = {
      Text.super.<init>();
      ()
    };
    final override <synthetic> def toString(): String = "Text";
    case <synthetic> def apply[T](value: T, source: String): Text[T] = new Text[T](value, source);
    case <synthetic> def unapply[T](x$0: Text[T]): Option[(T, String)] = if (x$0.==(null))
      scala.None
    else
      Some.apply[(T, String)](scala.Tuple2.apply[T, String](x$0.value, x$0.source));
    <synthetic> private def readResolve(): Object = Text
  }
```

We should probably enable mima first though.

Reply:
yeah, I considered doing that and wrote a few of them... then I figured I wanted to know if it was worth it first. Making it by-name introduces a small performance penalty (the extra object allocation) and I wanted to make sure @lihaoyi was ok with that. But since users already opt-in to the memory alloc overhead of `Text` in the first place, I thought that'd be ok.

The other alternative would be to leave `Text` alone, possibly deprecated until the next major scala release, and introduce a new `SourceText` beside it.